receiver/snmpreceiver/scraper.go

```diff
@@ -0,0 +1,559 @@
+// Copyright 2020 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snmpreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/snmpreceiver"
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-errors/errors"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+)
+
+var (
+	errScrape = errors.New("failed to successfully scrape any SNMP metrics")
+)
+
+var (
+	errMsgBadValueType                  = `returned metric SNMP data type for OID: %s is not supported`
+	errMsgIndexedBadValueType           = `returned metric SNMP data type for OID: %s from column OID: %s is not supported`
+	errMsgBadIndexedAttributes          = `problem retrieving SNMP indexed attribute data: %w`
+	errMsgBadIndexedResourceAttributes  = `problem retrieving SNMP indexed resource attribute data: %w`
+	errMsgIndexedAttributesBadValueType = `returned attribute SNMP data type for OID: %s from column OID: %s is not supported`
+)
+
+// snmpScraper handles scraping of SNMP metrics
+type snmpScraper struct {
+	client   client
+	logger   *zap.Logger
+	cfg      *Config
+	settings component.ReceiverCreateSettings
+}
+
+// newScraper creates an initialized snmpScraper
+func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCreateSettings) *snmpScraper {
+	return &snmpScraper{
+		logger:   logger,
+		cfg:      cfg,
+		settings: settings,
+	}
+}
+
+// start gets the client ready
+func (s *snmpScraper) start(ctx context.Context, host component.Host) (err error) {
+	s.client, err = newClient(s.cfg, host, s.settings.TelemetrySettings, s.logger)
+	if err != nil {
+		return err
+	}
+	err = s.client.Connect()
+
+	return
+}
+
+// shutdown closes the client
+func (s *snmpScraper) shutdown(ctx context.Context, host component.Host) (err error) {
```

Comment:
nit: This doesn't need a named return value. This function could just be

```go
func (s *snmpScraper) shutdown(ctx context.Context, host component.Host) error {
    return s.client.Close()
}
```

Reply:
changed